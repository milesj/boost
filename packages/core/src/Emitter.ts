import camelCase from 'lodash/camelCase';
import upperFirst from 'lodash/upperFirst';
import { EVENT_NAME_PATTERN } from './constants';

export type EventArguments = any[];

export type EventListener = (...args: EventArguments) => false | void;

export default class Emitter {
  listeners: Map<string, Set<EventListener>> = new Map();

  /**
   * Remove all listeners for the defined event name.
   */
  clearListeners(eventName: string): this {
    this.getListeners(eventName).clear();

    return this;
  }

  /**
   * Syncronously execute listeners for the defined event and arguments.
   */
  emit(eventName: string, args: EventArguments = []): this {
    Array.from(this.getListeners(eventName)).some(listener => listener(...args) === false);

    return this;
  }

  /**
   * Return all event names with registered listeners.
   */
  getEventNames(): string[] {
    return Array.from(this.listeners.keys());
  }

  /**
   * Return a set of listeners for a specific event name.
   */
  getListeners(eventName: string): Set<EventListener> {
    if (!eventName.match(EVENT_NAME_PATTERN)) {
      throw new Error(
        `Invalid event name "${eventName}". ` +
          'May only contain dashes, periods, and lowercase characters.',
      );
    }

    if (!this.listeners.has(eventName)) {
      this.listeners.set(eventName, new Set());
    }

    return this.listeners.get(eventName)!;
  }

  /**
   * Remove a listener function from a specific event name.
   */
  off(eventName: string, listener: EventListener): this {
    this.getListeners(eventName).delete(listener);

    return this;
  }

  /**
   * Register a listener function to a specific event name.
   */
  on(eventName: string, listener: EventListener): this {
    if (typeof listener !== 'function') {
      throw new TypeError(`Invalid event listener for "${eventName}", must be a function.`);
    }

    let eventProp = eventName;
    const args = ['listener'];

    if (eventName.includes('.')) {
      const [scope, event] = eventName.split('.', 2);

      args.push(`'${scope}'`);
      eventProp = event;
    }

    console.warn(
      `Boost emitter has been deprecated. Please migrate \`on('${eventName}', listener)\` to the new event system, \`on${upperFirst(
        camelCase(eventProp),
      )}.listen(${args.join(', ')})\`.'`,
    );

    this.getListeners(eventName).add(listener);

    return this;
  }
}
