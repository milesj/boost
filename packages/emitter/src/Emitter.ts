import { EVENT_NAME_PATTERN } from './constants';
import { ListenerType, Arguments, WaterfallArgument } from './types';

export default class Emitter<T> {
  listeners: { [K in keyof T]?: Set<ListenerType<T[K]>> } = {};

  /**
   * Remove all listeners for the defined event name.
   */
  clearListeners<K extends keyof T>(eventName: K): this {
    this.getListeners(eventName).clear();

    return this;
  }

  /**
   * Synchronously execute listeners for the defined event and arguments.
   */
  emit<K extends keyof T>(eventName: K, args: Arguments<T[K]>): this {
    Array.from(this.getListeners(eventName)).forEach(listener => {
      listener(...args);
    });

    return this;
  }

  /**
   * Synchronously execute listeners for the defined event and arguments.
   * If a listener returns `false`, the loop with be aborted early.
   */
  emitBail<K extends keyof T>(eventName: K, args: Arguments<T[K]>): this {
    Array.from(this.getListeners(eventName)).some(listener => listener(...args) === false);

    return this;
  }

  /**
   * Asynchronously execute listeners for the defined event and arguments.
   * Will return a promise with an array of each listener result.
   */
  emitParallel<K extends keyof T>(eventName: K, args: Arguments<T[K]>): Promise<any[]> {
    return Promise.all(
      Array.from(this.getListeners(eventName)).map(listener => Promise.resolve(listener(...args))),
    );
  }

  /**
   * Synchronously execute listeners for the defined event and value.
   * The return value of each listener will be passed as an argument to the next listener.
   */
  emitWaterfall<K extends keyof T>(
    eventName: K,
    value: WaterfallArgument<T[K]>,
  ): WaterfallArgument<T[K]> {
    return Array.from(this.getListeners(eventName)).reduce(
      (nextValue, listener) => listener(nextValue),
      value,
    );
  }

  /**
   * Return all event names with registered listeners.
   */
<<<<<<< HEAD:packages/core/src/Emitter.ts
  getEventNames(): string[] {
    return Array.from(this.listeners.keys());
=======
  getEventNames(): (keyof T)[] {
    return Object.keys(this.listeners) as (keyof T)[];
>>>>>>> a8bbc3b... Start moving emitter to own package.:packages/emitter/src/Emitter.ts
  }

  /**
   * Return a set of listeners for a specific event name.
   */
  getListeners<K extends keyof T>(eventName: K): Set<ListenerType<T[K]>> {
    const key = String(eventName);

    if (!key.match(EVENT_NAME_PATTERN)) {
      throw new Error(
        `Invalid event name "${eventName}". ` +
          'May only contain dashes, periods, and lowercase characters.',
      );
    }

    if (!this.listeners.has(eventName)) {
      this.listeners.set(eventName, new Set());
    }

<<<<<<< HEAD:packages/core/src/Emitter.ts
    return this.listeners.get(eventName)!;
=======
    return this.listeners[eventName]!;
>>>>>>> a8bbc3b... Start moving emitter to own package.:packages/emitter/src/Emitter.ts
  }

  /**
   * Remove a listener from a specific event name.
   */
  off<K extends keyof T>(eventName: K, listener: ListenerType<T[K]>): this {
    this.getListeners(eventName).delete(listener);

    return this;
  }

  /**
   * Register a listener to a specific event name.
   */
  on<K extends keyof T>(eventName: K, listener: ListenerType<T[K]>): this {
    this.getListeners(eventName).add(this.validateListener(eventName, listener));

    return this;
  }

  /**
   * Register a listener to a specific event name that only triggers once.
   */
  once<K extends keyof T>(eventName: K, listener: ListenerType<T[K]>): this {
    const func = this.validateListener(eventName, listener);
    const handler: any = (...args: Arguments<T[K]>) => {
      this.off(eventName, handler);

      return func(...args);
    };

    return this.on(eventName, handler);
  }

  protected validateListener<K extends keyof T, L>(eventName: K, listener: L): L {
    if (typeof listener !== 'function') {
      throw new TypeError(`Invalid event listener for "${eventName}", must be a function.`);
    }

    return listener;
  }
}
