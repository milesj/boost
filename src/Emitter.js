/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Event from './Event';
import { APP_NAME_PATTERN } from './constants';

import type { EventArguments, EventListener, Result } from './types';

export default class Emitter {
  listeners: { [eventName: string]: Set<EventListener> } = {};

  /**
   * Syncronously execute listeners for the defined event and arguments.
   */
  emit(eventName: string, initialValue: Result, args?: EventArguments = []): Event {
    const event = new Event(eventName, initialValue);

    Array.from(this.getListeners(event.name)).some((listener) => {
      listener(event, ...args);

      return event.stopped;
    });

    return event;
  }

  /**
   * Syncronously execute listeners for the defined event and arguments,
   * through a chaining lyer controlled by next handlers.
   */
  emitCascade(eventName: string, initialValue: Result, args?: EventArguments = []): Event {
    const event = new Event(eventName, initialValue);
    const listeners = Array.from(this.getListeners(event.name));
    let index = 0;

    if (listeners.length === 0) {
      return event;
    }

    // Handler passed to each listener
    function next(nextIndex: number, ...nextEventArguments: EventArguments) {
      if (nextIndex < index || nextIndex > listeners.length) {
        throw new Error('Invalid cascading event.');
      }

      index = nextIndex;
      const listener = listeners[index];

      if (!listener || event.stopped) {
        return;
      }

      // Set the next handler
      event.next = () => next(index + 1, ...nextEventArguments);

      listener(event, ...nextEventArguments);
    }

    // Execute the first listener
    next(0, ...args);

    return event;
  }

  /**
   * Return all event names with registered listeners.
   */
  getEventNames(): string[] {
    return Object.keys(this.listeners);
  }

  /**
   * Return a set of listeners for a specific event name.
   */
  getListeners(eventName: string): Set<EventListener> {
    if (!eventName.match(APP_NAME_PATTERN)) {
      throw new Error(
        `Invalid event name "${eventName}". ` +
        'May only contain dashes and lowercase characters.',
      );
    }

    if (!this.listeners[eventName]) {
      this.listeners[eventName] = new Set();
    }

    return this.listeners[eventName];
  }

  /**
   * Remove a listener function from a specific event name.
   */
  off(eventName: string, listener: EventListener): this {
    this.getListeners(eventName).delete(listener);

    return this;
  }

  /**
   * Register a listener function to a specific event name.
   */
  on(eventName: string, listener: EventListener): this {
    if (typeof listener !== 'function') {
      throw new TypeError(`Invalid event listener for "${eventName}", must be a function.`);
    }

    this.getListeners(eventName).add(listener);

    return this;
  }
}
