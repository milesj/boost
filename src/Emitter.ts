/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Event from './Event';
import { APP_NAME_PATTERN } from './constants';
import { EventArguments, EventListener } from './types';

export interface EmitterInterface {
  emit(name: string, args: EventArguments, initialValue: any): Event;
  off(eventName: string, listener: EventListener): this;
  on(eventName: string, listener: EventListener): this;
}

export default class Emitter implements EmitterInterface {
  listeners: { [eventName: string]: Set<EventListener> } = {};

  namespace: string = '';

  /**
   * Create an event name with optional namespace.
   */
  createEventName(name: string): string {
    if (this.namespace && !name.startsWith(this.namespace)) {
      return `${this.namespace}.${name}`;
    }

    return name;
  }

  /**
   * Syncronously execute listeners for the defined event and arguments.
   */
  emit(name: string, args: EventArguments = [], initialValue: any = null): Event {
    const event = new Event(this.createEventName(name), initialValue);

    Array.from(this.getListeners(event.name)).some((listener) => {
      listener(event, ...args);

      return event.stopped;
    });

    return event;
  }

  /**
   * Syncronously execute listeners for the defined event and arguments,
   * through a chaining lyer controlled by next handlers.
   */
  emitCascade(name: string, args: EventArguments = [], initialValue: any = null): Event {
    const event = new Event(this.createEventName(name), initialValue);
    const listeners = Array.from(this.getListeners(event.name));
    let index = 0;

    if (listeners.length === 0) {
      return event;
    }

    // Handler passed to each listener
    function next(nextIndex: number, ...nextEventArguments: EventArguments) {
      index = nextIndex;
      const listener = listeners[index];

      if (!listener || event.stopped) {
        return;
      }

      // Set the next handler
      event.next = () => next(index + 1, ...nextEventArguments);

      listener(event, ...nextEventArguments);
    }

    // Execute the first listener
    next(0, ...args);

    return event;
  }

  /**
   * Return all event names with registered listeners.
   */
  getEventNames(): string[] {
    return Object.keys(this.listeners);
  }

  /**
   * Return a set of listeners for a specific event name.
   */
  getListeners(eventName: string): Set<EventListener> {
    if (!eventName.match(APP_NAME_PATTERN)) {
      throw new Error(
        `Invalid event name "${eventName}". ` +
        'May only contain dashes, periods, and lowercase characters.',
      );
    }

    if (!this.listeners[eventName]) {
      this.listeners[eventName] = new Set();
    }

    return this.listeners[eventName];
  }

  /**
   * Remove a listener function from a specific event name.
   */
  off(eventName: string, listener: EventListener): this {
    this.getListeners(eventName).delete(listener);

    return this;
  }

  /**
   * Register a listener function to a specific event name.
   */
  on(eventName: string, listener: EventListener): this {
    if (typeof listener !== 'function') {
      throw new TypeError(`Invalid event listener for "${eventName}", must be a function.`);
    }

    this.getListeners(eventName).add(listener);

    return this;
  }

  /**
   * Set the namespace.
   */
  setEventNamespace(namespace: string): this {
    this.namespace = namespace;

    return this;
  }

  /**
   * Remove the namespace.
   */
  removeEventNamespace(): this {
    this.namespace = '';

    return this;
  }
}
