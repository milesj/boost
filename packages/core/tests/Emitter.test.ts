import Emitter from '../src/Emitter';

describe('Emitter', () => {
  let emitter: Emitter;

  beforeEach(() => {
    emitter = new Emitter();
  });

  it('toggles event namespace', () => {
    expect(emitter.namespace).toBe('');

    emitter.setEventNamespace('ns');

    expect(emitter.namespace).toBe('ns');

    emitter.removeEventNamespace();

    expect(emitter.namespace).toBe('');
  });

  describe('createEventName()', () => {
    it('returns name as-is if no namespace', () => {
      expect(emitter.createEventName('foo')).toBe('foo');
    });

    it('prepends namespace', () => {
      emitter.setEventNamespace('ns');

      expect(emitter.createEventName('foo')).toBe('ns.foo');
    });

    it('doesnt prepend namespace if already prefixed', () => {
      emitter.setEventNamespace('ns');

      expect(emitter.createEventName('ns.foo')).toBe('ns.foo');
    });
  });

  describe('emit()', () => {
    it('executes listeners in order', () => {
      let output = '';

      emitter.on('foo', () => {
        output += 'A';
      });
      emitter.on('foo', () => {
        output += 'B';
      });
      emitter.on('foo', () => {
        output += 'C';
      });
      emitter.emit('foo');

      expect(output).toBe('ABC');
    });

    it('executes listeners syncronously with arguments', () => {
      const output: number[] = [];

      emitter.on('foo', (a, b) => {
        output.push(a, b * 2);
      });
      emitter.on('foo', (a, b) => {
        output.push(a * 3, b * 4);
      });
      emitter.on('foo', (a, b) => {
        output.push(a * 5, b * 6);
      });
      emitter.emit('foo', [2, 3]);

      expect(output).toEqual([2, 6, 6, 12, 10, 18]);
    });

    it('executes listeners syncronously while passing values to each', () => {
      let value = 'foo';

      emitter.on('foo', () => {
        value = value.toUpperCase();
      });
      emitter.on('foo', () => {
        value = value
          .split('')
          .reverse()
          .join('');
      });
      emitter.on('foo', () => {
        value = `${value}-${value}`;
      });

      emitter.emit('foo');

      expect(value).toBe('OOF-OOF');
    });

    it('executes listeners syncronously with arguments while passing values to each', () => {
      const value: string[] = [];

      emitter.on('foo', (a, b, c) => {
        value.push(a.repeat(3));
      });
      emitter.on('foo', (a, b, c) => {
        value.push(b.repeat(2));
      });
      emitter.on('foo', (a, b, c) => {
        value.push(c.repeat(1));
      });

      emitter.emit('foo', ['foo', 'bar', 'baz']);

      expect(value).toEqual(['foofoofoo', 'barbar', 'baz']);
    });

    it('execution can be stopped', () => {
      let count = 0;

      emitter.on('foo', () => {
        count += 1;
      });
      emitter.on('foo', () => false);
      emitter.on('foo', () => {
        count += 1;
      });
      emitter.emit('foo');

      expect(count).toBe(1);
    });

    it('passes arguments to listeners', () => {
      const baseArgs = [1, 2, 3];
      let args;

      emitter.on('foo', (...eventArgs) => {
        args = eventArgs;
      });
      emitter.emit('foo', baseArgs);

      expect(args).toEqual(baseArgs);
    });

    it('passes value by modifying event object', () => {
      let value = 0;

      emitter.on('foo', () => {
        value += 1;
      });
      emitter.on('foo', () => {
        value += 1;
      });
      emitter.on('foo', () => {
        value += 1;
      });

      emitter.emit('foo');

      expect(value).toBe(3);
    });

    describe('with namespace', () => {
      beforeEach(() => {
        emitter.setEventNamespace('ns');
      });

      afterEach(() => {
        emitter.removeEventNamespace();
      });

      it('executes listeners in order', () => {
        let output = '';

        emitter.on('ns.foo', () => {
          output += 'A';
        });
        emitter.on('ns.foo', () => {
          output += 'B';
        });
        emitter.on('ns.foo', () => {
          output += 'C';
        });
        emitter.emit('foo');

        expect(output).toBe('ABC');
      });
    });
  });

  describe('getEventNames()', () => {
    it('returns an array of event names', () => {
      emitter.getListeners('foo');
      emitter.getListeners('bar');
      emitter.getListeners('baz');
      emitter.getListeners('ns.qux');

      expect(emitter.getEventNames()).toEqual(['foo', 'bar', 'baz', 'ns.qux']);
    });
  });

  describe('getListeners()', () => {
    it('errors if name contains invalid characters', () => {
      expect(() => emitter.getListeners('foo+bar')).toThrowError(
        'Invalid event name "foo+bar". May only contain dashes, periods, and lowercase characters.',
      );
    });

    it('creates the listeners set if it does not exist', () => {
      expect(emitter.listeners.foo).toBeUndefined();

      const set = emitter.getListeners('foo');

      expect(set).toBeInstanceOf(Set);
      expect(emitter.listeners.foo).toBeDefined();
    });
  });

  describe('off()', () => {
    it('removes a listener from the set', () => {
      const listener = () => {};

      emitter.on('foo', listener);

      expect(emitter.getListeners('foo').has(listener)).toBe(true);

      emitter.off('foo', listener);

      expect(emitter.getListeners('foo').has(listener)).toBe(false);
    });
  });

  describe('on()', () => {
    it('errors if listener is not a function', () => {
      expect(() => {
        // @ts-ignore
        emitter.on('foo', 123);
      }).toThrowError('Invalid event listener for "foo", must be a function.');
    });

    it('adds a listener to the set', () => {
      const listener = () => {};

      expect(emitter.getListeners('foo').has(listener)).toBe(false);

      emitter.on('foo', listener);

      expect(emitter.getListeners('foo').has(listener)).toBe(true);
    });
  });
});
