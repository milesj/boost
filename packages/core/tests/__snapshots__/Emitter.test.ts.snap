// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emitter getListeners() errors if name contains invalid characters 1`] = `"Invalid event name \\"foo+bar\\". May only contain dashes, periods, and lowercase characters."`;

exports[`Emitter on() errors if listener is not a function 1`] = `"Invalid event listener for \\"foo\\", must be a function."`;
